### PR TITLE
Add locking between replicaGCQueue and multiraft.state.createGroup.

### PR DESCRIPTION
**Pull request by bdarnell**
This partially addresses the race seen in #2815. A similar race still
occurs but much less frequently.

### CODE REVIEW

**Comment by bdarnell** [comment]
I'm not sure how to test this. I've been testing it locally by increasing the iteration count in `TestRaftRemoveRace`, although checking in that change would noticeably increase total test runtime. I don't see a good way to trigger the race in a more direct/controlled way.

File: storage/replica_gc_queue.go [hunk 1]
```diff
@@ -124,6 +127,20 @@ func (q *replicaGCQueue) process(now roachpb.Timestamp, rng *Replica, _ *config.
...
+		// Instead, we check after acquiring the lock to make sure the
+		// range is still absent.
+		q.locker.Lock()
+		defer q.locker.Unlock()
+
+		if _, err := rng.rm.GetReplica(desc.RangeID); err == nil {
+			log.Infof("replica recreated during deletion; aborting deletion")
+		}
+
 		// TODO(bdarnell): add some sort of locking to prevent the range
```
Comment (tamird) [inline storage/replica_gc_queue.go]: remove?

**Comment by tamird** [comment]
What is the similar race? can you document it?

**Comment by bdarnell** [comment]
I'm still working on identifying the similar race. All I know so far is that it produces the same error message as #2815 and it takes over a minute for the test to reproduce it.

**Comment by bdarnell** [comment]
One "similar race" is that I forgot to `return nil` after the "aborting deletion" log line. But even with that fixed I'm seeing other rare failures. I suspect that what may be happening is that the node is sometimes falling far enough behind that it is learning about multiple iterations of the add/remove loop at once.

File: storage/replica_gc_queue.go (hunk 1 above)
Comment (bdarnell) [inline storage/replica_gc_queue.go]: Done.

**Comment by tamird** [comment]
LGTM

**Pull request merged**